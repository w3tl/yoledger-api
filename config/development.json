{
  "credentials": {
    "algorithm": "dev"
  },
  "token": {
    "secret": "jwtsecret",
    "expiresIn": 3000
  }
}
