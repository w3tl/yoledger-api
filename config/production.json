{
  "credentials": {
    "algorithm": "argon2"
  },
  "port": "3030"
}
