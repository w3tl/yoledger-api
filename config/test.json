{
  "mongoUri": "mongodb://localhost:31000/?replicaSet=rs",
  "credentials": {
    "algorithm": "dev"
  },
  "token": {
    "secret": "jwtsecret",
    "expiresIn": 5
  }
}
