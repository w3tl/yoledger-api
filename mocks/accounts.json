[
  {
    "name": "Bank card",
    "type": "ASSET",
    "balance": 1500.5
  },
  {
    "name": "Cash",
    "type": "ASSET",
    "balance": 143
  },
  {
    "name": "Train",
    "type": "EXPENSE"
  },
  {
    "name": "Food",
    "type": "EXPENSE"
  }
]
