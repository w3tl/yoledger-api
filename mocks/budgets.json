[
  {
    "date": "2018-05-25",
    "account": {
      "name": "Food",
      "type": "EXPENSE"
    },
    "amount": 9,
    "balance": 0
  },
  {
    "date": "2018-05-25",
    "account": {
      "name": "Train",
      "type": "EXPENSE"
    },
    "amount": 29,
    "balance": 20
  },
  {
    "date": "2018-06-10",
    "account": {
      "name": "Food",
      "type": "EXPENSE"
    },
    "amount": 10,
    "balance": 0
  },
  {
    "date": "2018-06-10",
    "account": {
      "name": "Train",
      "type": "EXPENSE"
    },
    "amount": 30,
    "balance": 20
  },
  {
    "date": "2018-06-25",
    "account": {
      "name": "Food",
      "type": "EXPENSE"
    },
    "amount": 11,
    "balance": 1
  }
]
