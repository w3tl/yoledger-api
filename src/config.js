const convict = require('convict');

const config = convict({
  env: {
    doc: 'The application environment.',
    format: ['production', 'development', 'test', 'stage'],
    default: 'development',
    env: 'NODE_ENV',
  },
  port: {
    doc: 'The port to bind.',
    format: 'port',
    default: 3030,
    env: 'GRAPHQL_PORT',
    arg: 'port',
  },
  path: {
    doc: 'Custom path',
    env: 'GRAPHQL_PATH',
    default: 'graphql',
  },
  credentials: {
    algorithm: {
      doc: 'Current crypt algorithm',
      env: 'CRED_ALG',
      default: 'default',
    },
  },
  token: {
    secret: {
      doc: 'Secret used for jwt tokens',
      format: '*',
      default: '',
      sensitive: true,
      env: 'JWT_SECRET',
    },
    expiresIn: {
      doc: 'Expressed in seconds describing a time span',
      format: 'int',
      default: 60, // 1m
      env: 'TOKEN_EXPIRES',
    },
  },
  mongoUri: {
    doc: 'MongoDB connection string',
    format: '*',
    default: 'mongodb://localhost',
    env: 'MONGO_URI',
  },
});

// Load environment dependent configuration
const env = config.get('env');
config.loadFile(`config/${env}.json`);
// Perform validation
config.validate({ allowed: 'strict' });

module.exports = config;
