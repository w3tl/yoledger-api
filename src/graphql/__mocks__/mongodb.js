import {
  users,
  transactions as transactionsJson,
  accounts,
  budgets as budgetsJson,
} from '../mocks';

const mongodb = jest.genMockFromModule('mongodb');

const addCreatedAt = ({ createdAt, ...other }) => ({ ...other, createdAt: new Date(createdAt) });
const addDate = ({ date, ...other }) => ({ ...other, date: new Date(date) });
const transactions = transactionsJson.map(addCreatedAt);
const budgets = budgetsJson.map(addDate);

const mockSession = {
  startTransaction: () => Promise.resolve(),
  commitTransaction: () => Promise.resolve(),
  abortTransaction: () => Promise.reject({ // eslint-disable-line prefer-promise-reject-errors
    errorLabels: ['TransientTransactionError'],
  }),
};

const mockCollections = {
  users: {
    findOne: ({ _id }) => Promise.resolve(users.find(u => u._id === _id)),
    updateOne: () => Promise.resolve({ result: { nModified: 1 } }),
  },
  transactions: {
    find: () => ({
      skip: skip => ({
        limit: limit => ({
          toArray: () => transactions.slice(skip, skip + limit),
        }),
      }),
    }),
    findOne: ({ _id }) => Promise.resolve(transactions.find(t => t._id === _id)),
    insertOne: (transaction) => {
      transactions.push({ _id: '50', ...transaction });
      return Promise.resolve({ insertedId: '50' });
    },
    findOneAndDelete: ({ id }) => Promise.resolve({ value: transactions.find(t => t.id === id) }),
    updateOne: () => Promise.resolve(),
  },
  accounts: {
    find: query => ({
      toArray: () => {
        if ('type' in query) {
          return Promise.resolve(accounts.filter(acc => query.type === acc.type));
        }
        if ('name' in query) {
          if ('$in' in query.name) {
            const result = accounts.filter(({ name }) => query.name.$in.indexOf(name) !== -1);
            return Promise.resolve(result);
          }
        }
        return Promise.resolve([]);
      },
    }),
    findOne: ({ name, _id }) => Promise.resolve(
      accounts.find(acc => name === acc.name || _id === acc._id),
    ),
    insertOne: (account) => {
      accounts.push({ _id: '44', ...account });
      return Promise.resolve({ insertedId: '44' });
    },
    updateOne: () => Promise.resolve(),
  },
  budget: {
    find: ({ $and, date, ...other }) => ({
      toArray: () => {
        if ($and) {
          return Promise.resolve(
            budgets.filter(b => (
              b.account.name === other['account.name']
              && b.date.getTime() >= $and[0].date.$gte.getTime()
              && b.date.getTime() <= $and[1].date.$lte.getTime()
            )),
          );
        }
        return Promise.resolve(budgets.filter(b => b.date.getTime() === new Date(date).getTime()));
      },
    }),
    findOne: ({ account, date }) => {
      if (account) { return Promise.resolve(budgets.find(t => t.account.name === account)); }
      if (date) { return Promise.resolve(budgets.find(t => t.date.getTime() === date.getTime())); }
      return budgets[0];
    },
    updateOne: ({ date, account }, { $set }) => {
      const idx = budgets
        .findIndex(b => b.date.getTime() === date.getTime() && b.account.name === account.name);
      if (idx) {
        budgets[idx] = { ...budgets[idx], ...$set.modificator };
      }
      return Promise.resolve({ upsertedId: budgets[idx]._id });
    },
    aggregate: () => ({
      toArray: () => [
        accounts.find(a => a.name === 'Food'),
        accounts.find(a => a.name === 'Train'),
      ],
    }),
  },
};


mongodb.connection = {
  db: () => ({
    collection: name => mockCollections[name],
  }),
  withSession: callback => callback(mockSession),
};

mongodb.ObjectId = id => String(id);

module.exports = mongodb;
