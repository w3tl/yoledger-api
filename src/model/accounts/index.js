import Model from '../Model';

export const accountTypes = ['ASSET', 'EXPENSE', 'INCOME', 'VENDOR'];

const schema = {
  type: 'object',
  required: ['name', 'type', 'userId'],
  additionalProperties: false,
  properties: {
    _id: {
      bsonType: 'objectId',
    },
    name: {
      type: 'string',
    },
    balance: {
      type: 'number',
    },
    type: {
      enum: accountTypes,
    },
    userId: {
      type: 'string',
    },
  },
};

export default class Account extends Model {
  constructor(db, userId) {
    super(db, userId);
    this.schema = schema;
    this.collName = 'accounts';
  }

  create({
    name, type, balance = 0,
  }) {
    return super.insertOne({
      name,
      type,
      balance,
    });
  }

  addAmount(name, amount, session) {
    return super.updateOne(
      { name, userId: this.userId },
      { $inc: { balance: amount } },
      { session },
    );
  }

  findByName(name, ...props) {
    return super.findOne({ name }, ...props);
  }
}
