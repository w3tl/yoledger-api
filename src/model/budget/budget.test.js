import { MongoClient, ObjectId } from 'mongodb';
import config from '../../config';
import Budget from './index';
import Account from '../accounts';
import budgets from '../../../mocks/budgets';
import accounts from '../../../mocks/accounts';

let db;
let connection;
let budgetModel;

beforeAll(async () => {
  connection = await MongoClient.connect(config.get('mongoUri'), {
    connectWithNoPrimary: true,
    useNewUrlParser: true,
  });
  db = connection.db('dbbudgets');
  budgetModel = new Budget(db, 'admin');
  await budgetModel.init();
  const accountModel = new Account(db, 'admin');
  await accountModel.init();
  accountModel.insertMany(accounts);
});

afterAll(async () => {
  await connection.close();
});

describe('budget', () => {
  describe('upsertBudget', () => {
    beforeAll(async () => {
      await budgetModel.clear();
      await budgetModel
        .insertMany(
          budgets.map(({ date, ...other }) => ({ date: new Date(date), ...other })),
        );
    });

    afterAll(async () => {
      await budgetModel.clear();
      await budgetModel
        .insertMany(
          budgets.map(({ date, ...other }) => ({ date: new Date(date), ...other })),
        );
    });

    test('should update the amount in the existing period', async () => {
      const { date, account } = budgets[0];
      await budgetModel.upsertBudget({
        account: account.name,
        date,
        amount: 14,
      });
      const doc = await budgetModel.findOne({ date: new Date(date) });
      expect(doc).toMatchSnapshot({
        _id: expect.any(ObjectId),
      }, `should contain "${account}" account with amount of 14`);
    });

    test('should add a budget to the existing period', async () => {
      const { date } = budgets[0];
      await budgetModel.upsertBudget({
        account: 'Food',
        date,
        amount: 15,
      });
      const doc = await budgetModel.findOne({ date: new Date(date) });
      expect(doc).toMatchSnapshot({
        _id: expect.any(ObjectId),
      }, 'should contain "twix" account with amount of 15');
    });

    test('should create a non-existent period with a budget', async () => {
      const date = new Date('2018-01-13');
      await budgetModel.upsertBudget({
        account: 'Train',
        date,
        amount: 15,
      });
      const doc = await budgetModel.findOne({ date });
      expect(doc).toMatchSnapshot({
        _id: expect.any(ObjectId),
      }, 'should contain "twix" account with amount of 15');
    });
  });
});
