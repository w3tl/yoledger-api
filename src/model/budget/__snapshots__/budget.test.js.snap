// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`budget upsertBudget should add a budget to the existing period: should contain "twix" account with amount of 15 1`] = `
Object {
  "_id": Any<ObjectID>,
  "account": Object {
    "name": "Food",
    "type": "EXPENSE",
  },
  "amount": 15,
  "balance": 0,
  "date": 2018-05-25T00:00:00.000Z,
  "userId": "admin",
}
`;

exports[`budget upsertBudget should create a non-existent period with a budget: should contain "twix" account with amount of 15 1`] = `
Object {
  "_id": Any<ObjectID>,
  "account": Object {
    "name": "Train",
    "type": "EXPENSE",
  },
  "amount": 15,
  "date": 2018-01-13T00:00:00.000Z,
  "userId": "admin",
}
`;

exports[`budget upsertBudget should update the amount in the existing period: should contain "[object Object]" account with amount of 14 1`] = `
Object {
  "_id": Any<ObjectID>,
  "account": Object {
    "name": "Food",
    "type": "EXPENSE",
  },
  "amount": 14,
  "balance": 0,
  "date": 2018-05-25T00:00:00.000Z,
  "userId": "admin",
}
`;
